hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/utils/RuntimeBytecodeExtractor.java

```
@@ -53,17 +55,37 @@ private static String getRuntimeBytecode(final String initBytecode) {
         int codeCopyIndex = initBytecode.indexOf(CODECOPY);
 
         if (codeCopyIndex == -1) {
-            throw new RuntimeException("CODECOPY instruction (39) not found in init bytecode.");
+            throw new IllegalArgumentException("CODECOPY instruction (39) not found in init bytecode.");
         }
 
         // Find the first occurrence of "6080" after the "CODECOPY"
         int runtimeCodePrefixIndex = initBytecode.indexOf(RUNTIME_CODE_PREFIX, codeCopyIndex);
 
         if (runtimeCodePrefixIndex == -1) {
-            throw new RuntimeException("Runtime code prefix (6080) not found after CODECOPY.");
+            throw new IllegalArgumentException("Runtime code prefix (6080) not found after CODECOPY.");
         }
 
         // Extract the runtime bytecode starting from the runtimeCodePrefixIndex
         return initBytecode.substring(runtimeCodePrefixIndex);
     }
+
+    /**
+     * Checks if a given data string is likely init bytecode.
+     *
+     * @param data the data string to check.
+     * @return true if it is init bytecode, false otherwise.
+     */
+    public static boolean isInitBytecode(final String data) {
+        if (data == null || data.isEmpty()) {
+            return false;
+        }
```

We can add additional logic checking if the data starts with `606060` or `608060` sequence, as this is the current format, where PUSH1 operation is inserting a free memory pointer. In this way we can fail fast, if the `data` doesn't even begin with this sequence. I believe in a lot of cases this check might be very efficient and enough to distinguish between a `bytecode` and `calldata`.

---

This is a great suggestion -> however doing that there was another corner case here you can observe one of the autogenerated solidity classes EthCall.sol init code starts with 6104d25. I'm not sure what this case is but perhaps we should handle it as well. I'm now thinking if there can be other cases like that which we might miss.

---

I found [a great article](https://medium.com/coinmonks/evm-part-ii-the-journey-of-smart-contracts-from-solidity-code-to-bytecode-a21300b23dee) that explains the structure of the init bytecode. It seems that in each init bytecode there is the sequence "60806040" which is responsible for assigning a free memory pointer. Before that there could be some other instructions for constructing the contract. In your example for `EthCall.sol` the bytecode starts with "6104d2" and this means: "PUSH2 0x4D2", which is "push 1234" which is one of the fields in the contract. And also in each contract there is a `codecopy` instruction afterwards followed by the runtime bytecode. I think this method can be enhanced with these findings and also the position of the opcodes is important. For example - to contain "f3" for the `return` opcode after the `codecopy` and not anywhere in the string. The `codecopy` should be after the free memory pointer "60806040" and so on.

---

Thanks. Looks good now.